source/agora/node/Validator.d

```diff
@@ -553,8 +557,27 @@ public class Validator : FullNode, API
         {
             log.trace("Sending Enrollment at height {} for {} cycles with {}",
                 block_height, this.params.ValidatorCycle, enroll_key);
-            this.network.peers.each!(p => p.client.sendEnrollment(
-                this.enroll_man.createEnrollment(enroll_key, block_height + 1)));
+            const enrollment = this.enroll_man.createEnrollment(enroll_key,
+                block_height + 1);
+
+            // Disable existing timer
+            if (emitter)
+                emitter.stop();
+
+            // Network tests use 1 second block interval, timeout should be
+            // adjusted accordingly
+            version (unittest) enum EnrollmentDelay = 10.msecs;
+            else enum EnrollmentDelay = 10.seconds;
+
+            // This timer will fire after EnrollmentDelay. Within that period,
+            // if we decide to request a different enrollment; next call to
+            // `checkAndEnroll` will disable this. This way we wont flood the
+            // network with invalid enrollment requests
+            () @trusted {
+                emitter = this.taskman.setTimer(dont_wait ? 0.msecs : EnrollmentDelay,
+                    {this.network.peers.each!(p => p.client.sendEnrollment(enrollment));}, 
+                    Periodic.No);
+            } ();
```

Comment:
So you essentially want to turn this into a "daemon" that gets periodically called and checks if it needs to enroll, right ? In which case taking the timer logic out of the function might be the best approach.

Reply:
Not really, it is a one shot timer. What I wanted is essentially a delayed and cancelable task.